src/plugins/src/infra_tester_plugins/assertion.py

```diff
@@ -0,0 +1,96 @@
+        name of the input and the value being the value of the input.
+
+        This method should return None if the inputs are valid.
+        Otherwise, it should return a string describing the error.
+
+        Any exception thrown by this method will be treated as an
+        implementation error and will be logged as such.
+
+        Args:
+            inputs (dict[str, object]): The inputs provided to the
+            plugin.
+
+        Raises:
+            NotImplementedError: If the plugin does not implement
+            this method.
+
+        Returns:
+            Union[str, None]: If the inputs are valid, return None.
+            Otherwise, return a string describing the error.
+        """
+
+        # It's better to encourage input validation by the plugin
+        # developers by forcing them to implement this method.
+
+        raise NotImplementedError(
+            f"Plugin must implement the method \
+              '{sys._getframe().f_code.co_name}'"
+        )
+
+    def run_assertion(
+        self, inputs: dict[Any, Any], state: dict[Any, Any]
+    ) -> Union[str, None]:
+        """
+        This method should contain the logic to run the assertion
+        and return the result.
+
+        If the assertion fails, this method should return a
+        string describing the error. Otherwise, it should return
+        None.
+
+        This method should not raise any exception. Any exception
+        thrown by this method will be treated as an implementation
+        error and will be logged as such.
+
+        Args:
+            inputs (dict[str, object]): The inputs provided to the
+            plugin.
+
+            state (dict[str, object]): The current Terraform state
+            as a dictionary.
+
+        Raises:
+            NotImplementedError: If the plugin does not implement
+            this method.
+
+        Returns:
+            Union[str, None]: None if the assertion passes. Otherwise
+            return a string describing why the assertion failed.
+        """
+
+        raise NotImplementedError(
+            f"Plugin must implement the method \
+             '{sys._getframe().f_code.co_name}'"
+        )
+
+    def cleanup(self, inputs: dict[Any, Any], state: dict[Any, Any]):
+        """
+        Cleanup any resources used by the plugin. This method
+        will be called after the plugin has been run regardless
+        of whether the run was successful or not. It is optional
+        and does not need to be implemented if there's nothing to
+        clean up.
+
+        If this method is implemented, it should be
+        idempotent, i.e, cleanup should always have the same end
+        result regardless of how many times it is called. Cleanup
+        is considered successful if it does not raise any exception.
+
+        Args:
+            inputs (dict[str, object]): The inputs provided to the
+            plugin.
+
+            state (dict[str, object]): The current Terraform state
```

Comment:
Good catch, I changed the type but forgot to update the docs. I've fixed this in the latest changes. I'm also using the `Dict` object from `typing` instead of directly using the `dict` type so that the code will work on any version greater than python 3.8.